### PR TITLE
Allow sets and enforce uniqueness in string choices

### PR DESCRIPTION
**Pull request by fwkoch**
- StringChoice property attribute `choices` may now be a set in addition to list, tuple, or dict.
- If `choices` is a list or tuple, a `TypeError` will be raised if there are any repeated entries.
- If `choices` is a dictionary, a `TypeError` will be raised if there are any duplicate entries across all keys and values.
- Also dictionary values may now be tuples or sets in addition to lists

See: https://github.com/3ptscience/properties/issues/51

### CODE REVIEW

**Comment by fwkoch** [comment]
@bsmithyman - Thanks for the review. Regarding the sub-items, I agree `list`s are not exceptional. I'm inclined to just leave it as is for now though for a few reasons:
1. The `choices` dictionary should be so small that speed is irrelevant.
2. I think it should error if there are duplicate choice values rather than silently coercing to `set`/`frozenset` (so we'd still need to explicitly check for uniqueness).
3. Even if the sub-items are `frozenset`s I think there would still be additional messy logic ensuring uniqueness across keys and all sub-items.
If more complex use-cases of `StringChoice` emerge, this should definitely be revisited.

**Comment by bsmithyman** [comment]
@fwkoch: Makes sense to me. I am fine w/ this; all looks good.

File: properties/basic.py [hunk 1]
```diff
@@ -573,19 +573,27 @@ def choices(self):
...
-            if not isinstance(val, (list, tuple)):
+            if isinstance(val, (set, list, tuple)):
+                value[key] = list(val)
+            else:
                 value[key] = [val]
+        all_items = []
         for key, val in value.items():
             if not isinstance(key, string_types):
-                raise TypeError('value must be strings')
+                raise TypeError("'choices' must be strings'")
```
Comment (ThomasMatern) [inline properties/basic.py]: Just one tiny remark - the quotation mark after 'strings' shouldn't be there.

File: properties/basic.py [hunk 2]
```diff
@@ -573,19 +573,27 @@ def choices(self):
...
             for sub_val in val:
                 if not isinstance(sub_val, string_types):
-                    raise TypeError('value must be strings')
+                    raise TypeError("'choices' must be strings")
+            all_items = all_items + [key] + val
```
Comment (ThomasMatern) [inline properties/basic.py]: I think `all_items += [key] + val ` is easier to read and faster (does not matter too much here)

Comment (fwkoch) [inline properties/basic.py]: I agree.

**Pull request merged**